File: src/lib/portfolioEngine.ts
import type {
  CitationGapResult,
  RecurringGapInsight,
  PortfolioAnalysisResult,
  PerPromptBreakdown,
} from "./types";

export class PortfolioAggregator {
  private static WEIGHT_MULTIPLIERS: Record<string, number> = {
    CRITICAL: 3.0,
    HIGH: 2.0,
    MEDIUM: 1.0,
  };

  static aggregate(
    brandDomain: string,
    results: CitationGapResult[],
    topN: number = 10
  ): PortfolioAnalysisResult {
    if (!results || results.length === 0) {
      return {
        brandDomain,
        totalPromptsAnalyzed: 0,
        totalDistinctCompetitors: 0,
        prompts: [],
        perPromptBreakdowns: [],
        recurringGaps: [],
        bulkMarkdownBrief: `# Peec AI Portfolio Remediation Brief\n\nNo prompts analyzed.`,
        sprintJiraBacklog: `# Jira Sprint Backlog\n\nNo issues generated.`,
        timestamp: new Date().toISOString(),
      };
    }

    const totalPrompts = results.length;
    const distinctCompetitors = new Set<string>();
    const promptsList: string[] = [];
    const perPromptBreakdowns: PerPromptBreakdown[] = [];

    // Accumulator map: gapKey -> aggregated object
    const accumulator: Map<
      string,
      {
        gapKey: string;
        gapType: 'Schema Markup' | 'Benchmark / Metric' | 'Topic Entity';
        displayName: string;
        citationWeight: 'CRITICAL' | 'HIGH' | 'MEDIUM';
        prompts: Set<string>;
        competitorUrls: Set<string>;
        recs: string[];
        codeSnippet: string;
        lossReason: string;
      }
    > = new Map();

    for (const res of results) {
      promptsList.push(res.query);
      const compDomain = res.competitorPayload.domain || "competitor.com";
      distinctCompetitors.add(compDomain);
      const compUrl = res.competitorPayload.url;

      perPromptBreakdowns.push({
        query: res.query,
        competitorUrl: compUrl,
        competitorDomain: compDomain,
        schemaGapsCount: res.schemaGaps.length,
        benchmarkGapsCount: res.benchmarkGaps.length,
        entityGapsCount: res.entityGaps.length,
        topMissingSchema: res.schemaGaps.length > 0 ? res.schemaGaps[0].schemaType : "None",
        status: (res.schemaGaps.length > 0 || res.benchmarkGaps.length > 0) ? "CRITICAL_GAPS" : "AUDITED"
      });

      // 1. Process Schema Gaps
      for (const sg of res.schemaGaps) {
        const key = `schema:${sg.schemaType}`;
        if (!accumulator.has(key)) {
          accumulator.set(key, {
            gapKey: key,
            gapType: "Schema Markup",
            displayName: `Schema.org @type ${sg.schemaType}`,
            citationWeight: "CRITICAL",
            prompts: new Set(),
            competitorUrls: new Set(),
            recs: [],
            codeSnippet: sg.recommendedJsonLd || `<script type="application/ld+json">\n{\n  "@context": "https://schema.org",\n  "@type": "${sg.schemaType}"\n}\n</script>`,
            lossReason: sg.impactReason || "Parsed directly by AI search engines (Perplexity, ChatGPT Search) to ground structured answer cards."
          });
        }
        const acc = accumulator.get(key)!;
        acc.prompts.add(res.query);
        acc.competitorUrls.add(compUrl);
        if (sg.impactReason && !acc.recs.includes(sg.impactReason)) {
          acc.recs.push(sg.impactReason);
        }
        if (sg.recommendedJsonLd && acc.codeSnippet.length < sg.recommendedJsonLd.length) {
          acc.codeSnippet = sg.recommendedJsonLd;
        }
      }

      // 2. Process Benchmark Gaps
      for (const bg of res.benchmarkGaps) {
        const key = `benchmark:${bg.metricName.toLowerCase().trim()}`;
        if (!accumulator.has(key)) {
          accumulator.set(key, {
            gapKey: key,
            gapType: "Benchmark / Metric",
            displayName: `Quantitative Metric: ${bg.metricName}`,
            citationWeight: "HIGH",
            prompts: new Set(),
            competitorUrls: new Set(),
            recs: [],
            codeSnippet: `// Suggested landing page stat insertion:\n"${bg.metricName}: ${bg.competitorValue}"\nRecommendation: ${bg.recommendation}`,
            lossReason: `Competitors explicitly quote "${bg.competitorValue}", giving LLMs hard numerical evidence for comparison queries.`
          });
        }
        const acc = accumulator.get(key)!;
        acc.prompts.add(res.query);
        acc.competitorUrls.add(compUrl);
        if (bg.recommendation && !acc.recs.includes(bg.recommendation)) {
          acc.recs.push(bg.recommendation);
        }
      }

      // 3. Process Entity Gaps
      for (const eg of res.entityGaps) {
        const key = `entity:${eg.entityName.toLowerCase().trim()}`;
        const weight = eg.citationWeight || "HIGH";
        if (!accumulator.has(key)) {
          accumulator.set(key, {
            gapKey: key,
            gapType: "Topic Entity",
            displayName: `Topic Entity: ${eg.entityName}`,
            citationWeight: weight,
            prompts: new Set(),
            competitorUrls: new Set(),
            recs: [],
            codeSnippet: `<!-- Section Insertion Recommendation: -->\n<h3>${eg.entityName}</h3>\n<p>${eg.actionPlan}</p>`,
            lossReason: eg.searchRelevance || "High search frequency in AI query context graphs."
          });
        }
        const acc = accumulator.get(key)!;
        acc.prompts.add(res.query);
        acc.competitorUrls.add(compUrl);
        if (eg.actionPlan && !acc.recs.includes(eg.actionPlan)) {
          acc.recs.push(eg.actionPlan);
        }
      }
    }

    const insights: RecurringGapInsight[] = [];
    for (const [key, data] of accumulator.entries()) {
      const recurrence = data.prompts.size;
      const weight = data.citationWeight;
      const multiplier = this.WEIGHT_MULTIPLIERS[weight] || 1.0;
      const score = Math.round(recurrence * multiplier * 10) / 10;
      const repRec = data.recs.length > 0 ? data.recs.slice(0, 2).join("; ") : "Add missing content or schema markup.";

      insights.push({
        gapKey: key,
        gapType: data.gapType,
        displayName: data.displayName,
        citationWeight: weight,
        recurrenceCount: recurrence,
        totalPromptsAnalyzed: totalPrompts,
        affectedPrompts: Array.from(data.prompts).sort(),
        exampleCompetitorUrls: Array.from(data.competitorUrls).sort(),
        representativeRecommendation: repRec,
        readyCodeSnippet: data.codeSnippet,
        lossReasonSummary: data.lossReason,
        priorityScore: score,
      });
    }

    // Sort descending by priorityScore, then recurrenceCount
    insights.sort((a, b) => {
      if (b.priorityScore !== a.priorityScore) {
        return b.priorityScore - a.priorityScore;
      }
      return b.recurrenceCount - a.recurrenceCount;
    });

    const topInsights = insights.slice(0, topN);
    const bulkBrief = this.generateBulkBrief(
      brandDomain,
      totalPrompts,
      distinctCompetitors.size,
      topInsights,
      perPromptBreakdowns
    );

    const jiraBacklog = this.generateJiraBacklog(
      brandDomain,
      topInsights
    );

    return {
      brandDomain,
      totalPromptsAnalyzed: totalPrompts,
      totalDistinctCompetitors: distinctCompetitors.size,
      prompts: promptsList,
      perPromptBreakdowns,
      recurringGaps: topInsights,
      bulkMarkdownBrief: bulkBrief,
      sprintJiraBacklog: jiraBacklog,
      timestamp: new Date().toISOString(),
    };
  }

  private static generateBulkBrief(
    brandDomain: string,
    totalPrompts: number,
    totalComps: number,
    insights: RecurringGapInsight[],
    breakdowns: PerPromptBreakdown[]
  ): string {
    let md = `# 📦 Peec AI Portfolio Remediation Brief: ${brandDomain}\n\n`;
    md += `* **Target Brand:** \`${brandDomain}\`\n`;
    md += `* **Monitored Prompts Analyzed:** \`${totalPrompts}\`\n`;
    md += `* **Distinct Winning Competitors:** \`${totalComps}\`\n`;
    md += `* **Top Recurring Fixes Identified:** \`${insights.length}\`\n\n`;
    md += `---\n\n## 📌 Executive Summary (High-Leverage Fixes)\n\n`;
    md += `Across ${totalPrompts} monitored AI search prompts where \`${brandDomain}\` is losing citations, the following recurring gaps represent the highest-leverage remediation actions. Fixing these core items once on your main landing page resolves citation deficits across multiple prompt clusters.\n\n`;

    md += `## 📋 Per-Prompt Audit Summary\n\n`;
    md += `| # | Monitored Prompt Query | Winning Competitor | Missing Schemas | Status |\n`;
    md += `|---|---|---|---|---|\n`;
    breakdowns.forEach((b, i) => {
      md += `| ${i + 1} | *"${b.query}"* | \`${b.competitorDomain}\` | \`${b.topMissingSchema}\` | ${b.status} |\n`;
    });
    md += `\n---\n\n`;

    md += `## 🎯 Ranked Recurring Fixes (Fix This Once, Win Multiple Prompts)\n\n`;
    insights.forEach((g, idx) => {
      const pct = g.totalPromptsAnalyzed > 0 ? Math.round((100 * g.recurrenceCount) / g.totalPromptsAnalyzed) : 0;
      md += `### #${idx + 1} · ${g.displayName} (\`${g.citationWeight}\` Priority)\n\n`;
      md += `* **Recurrence:** Appears on **${g.recurrenceCount}/${g.totalPromptsAnalyzed} prompts (${pct}%)**\n`;
      md += `* **Priority Score:** \`${g.priorityScore}\`\n`;
      md += `* **Why LLMs Favor Competitors:** ${g.lossReasonSummary}\n`;
      md += `* **Affected Prompts:**\n`;
      g.affectedPrompts.forEach((p) => {
        md += `  - *"${p}"*\n`;
      });
      md += `* **Seen on Competitor Pages:** ${g.exampleCompetitorUrls.map((u) => `[${u}](${u})`).join(", ")}\n`;
      md += `* **Action Required:** ${g.representativeRecommendation}\n\n`;
      md += `**Code / Content Remediation Snippet:**\n\`\`\`html\n${g.readyCodeSnippet}\n\`\`\`\n\n`;
    });

    return md;
  }

  private static generateJiraBacklog(
    brandDomain: string,
    insights: RecurringGapInsight[]
  ): string {
    let md = `h1. Peec AI Remediation Sprint Backlog: ${brandDomain}\n\n`;
    md += `*Epic:* PEEC-EPIC-12 (AI Search Citation Actionability)\n`;
    md += `*Target Brand:* ${brandDomain}\n\n`;

    insights.slice(0, 4).forEach((g, idx) => {
      const key = `PEEC-50${idx + 1}`;
      const points = g.citationWeight === "CRITICAL" ? 5 : (g.citationWeight === "HIGH" ? 3 : 2);
      md += `h2. [${key}] Fix Recurring Gap: ${g.displayName} (${points} Story Points)\n\n`;
      md += `*Issue Type:* Story\n`;
      md += `*Priority:* ${g.citationWeight === "CRITICAL" ? "Highest" : "High"}\n`;
      md += `*Recurrence:* ${g.recurrenceCount}/${g.totalPromptsAnalyzed} prompts\n\n`;
      md += `*User Story:*\n`;
      md += `AS A Growth Marketer / SEO Engineer,\n`;
      md += `I WANT TO implement ${g.displayName} on ${brandDomain},\n`;
      md += `SO THAT our brand wins citations across ${g.recurrenceCount} target AI search queries.\n\n`;
      md += `*Acceptance Criteria (Gherkin):*\n`;
      md += `{code}\n`;
      md += `Given the ${brandDomain} landing page is deployed\n`;
      md += `When Perplexity or ChatGPT crawls the HTML\n`;
      md += `Then ${g.displayName} is explicitly verified in the DOM\n`;
      md += `And citation eligibility increases across monitored prompt clusters.\n`;
      md += `{code}\n\n`;
    });

    return md;
  }
}

File: src/engine/portfolio_engine.py
"""
Portfolio Aggregator for Peec AI Citation Gap Engine.
Aggregates N independent CitationGapResults to identify recurring gap patterns across prompts.
"""

from typing import List, Dict
from src.models.schemas import (
    CitationGapResult,
    RecurringGapInsight,
    PortfolioAnalysisResult,
)


class PortfolioAggregator:
    """
    Rolls up N independent CitationGapResults, groups gaps by unique stable keys,
    computes priority scores based on recurrence across prompts and citation weights,
    and produces a comprehensive bulk remediation markdown brief.
    """

    WEIGHT_MULTIPLIERS = {
        "CRITICAL": 3.0,
        "HIGH": 2.0,
        "MEDIUM": 1.0,
    }

    @classmethod
    def aggregate(
        cls,
        brand_domain: str,
        results: List[CitationGapResult],
        top_n: int = 10,
    ) -> PortfolioAnalysisResult:
        if not results:
            return PortfolioAnalysisResult(
                brand_domain=brand_domain,
                total_prompts_analyzed=0,
                total_distinct_competitors=0,
                prompts=[],
                recurring_gaps=[],
                bulk_markdown_brief="# Peec AI Portfolio Remediation Brief\n\nNo prompts analyzed.",
            )

        total_prompts = len(results)
        distinct_competitors = set()
        prompts_list = []

        # Key -> Aggregation accumulator
        # gap_key: { 'gap_type', 'display_name', 'citation_weight', 'prompts': set(), 'competitor_urls': set(), 'recs': [] }
        accumulator: Dict[str, Dict] = {}

        for res in results:
            prompts_list.append(res.query)
            comp_domain = res.competitor_payload.domain
            distinct_competitors.add(comp_domain)
            comp_url = res.competitor_payload.url

            # 1. Process Schema Gaps
            for sg in res.schema_gaps:
                key = f"schema:{sg.schema_type}"
                if key not in accumulator:
                    accumulator[key] = {
                        "gap_key": key,
                        "gap_type": "Schema Markup",
                        "display_name": f"Schema @type {sg.schema_type}",
                        "citation_weight": "CRITICAL",
                        "prompts": set(),
                        "competitor_urls": set(),
                        "recs": [],
                    }
                accumulator[key]["prompts"].add(res.query)
                accumulator[key]["competitor_urls"].add(comp_url)
                if sg.impact_reason and sg.impact_reason not in accumulator[key]["recs"]:
                    accumulator[key]["recs"].append(sg.impact_reason)

            # 2. Process Benchmark Gaps
            for bg in res.benchmark_gaps:
                key = f"benchmark:{bg.metric_name.lower().strip()}"
                if key not in accumulator:
                    accumulator[key] = {
                        "gap_key": key,
                        "gap_type": "Benchmark / Metric",
                        "display_name": f"Metric: {bg.metric_name}",
                        "citation_weight": "HIGH",
                        "prompts": set(),
                        "competitor_urls": set(),
                        "recs": [],
                    }
                accumulator[key]["prompts"].add(res.query)
                accumulator[key]["competitor_urls"].add(comp_url)
                if bg.recommendation and bg.recommendation not in accumulator[key]["recs"]:
                    accumulator[key]["recs"].append(bg.recommendation)

            # 3. Process Entity Gaps
            for eg in res.entity_gaps:
                key = f"entity:{eg.entity_name.lower().strip()}"
                weight = eg.citation_weight or "HIGH"
                if key not in accumulator:
                    accumulator[key] = {
                        "gap_key": key,
                        "gap_type": "Topic Entity",
                        "display_name": f"Entity: {eg.entity_name}",
                        "citation_weight": weight,
                        "prompts": set(),
                        "competitor_urls": set(),
                        "recs": [],
                    }
                accumulator[key]["prompts"].add(res.query)
                accumulator[key]["competitor_urls"].add(comp_url)
                if eg.action_plan and eg.action_plan not in accumulator[key]["recs"]:
                    accumulator[key]["recs"].append(eg.action_plan)

        insights: List[RecurringGapInsight] = []
        for key, data in accumulator.items():
            recurrence = len(data["prompts"])
            weight = data["citation_weight"]
            multiplier = cls.WEIGHT_MULTIPLIERS.get(weight, 1.0)
            score = round(recurrence * multiplier, 2)

            rep_rec = "; ".join(data["recs"][:2]) if data["recs"] else "Add missing content or schema definition."

            insights.append(
                RecurringGapInsight(
                    gap_key=key,
                    gap_type=data["gap_type"],
                    display_name=data["display_name"],
                    citation_weight=weight,
                    recurrence_count=recurrence,
                    total_prompts_analyzed=total_prompts,
                    affected_prompts=sorted(list(data["prompts"])),
                    example_competitor_urls=sorted(list(data["competitor_urls"])),
                    representative_recommendation=rep_rec,
                    priority_score=score,
                )
            )

        # Sort descending by priority_score, then recurrence_count
        insights.sort(key=lambda x: (x.priority_score, x.recurrence_count), reverse=True)
        insights = insights[:top_n]

        bulk_brief = cls._generate_bulk_brief(brand_domain, total_prompts, len(distinct_competitors), insights)

        return PortfolioAnalysisResult(
            brand_domain=brand_domain,
            total_prompts_analyzed=total_prompts,
            total_distinct_competitors=len(distinct_competitors),
            prompts=prompts_list,
            recurring_gaps=insights,
            bulk_markdown_brief=bulk_brief,
        )

    @classmethod
    def _generate_bulk_brief(
        cls,
        brand_domain: str,
        total_prompts: int,
        total_comps: int,
        insights: List[RecurringGapInsight],
    ) -> str:
        md = f"# 📦 Peec AI Portfolio Remediation Brief: {brand_domain}\n\n"
        md += f"* **Target Brand:** `{brand_domain}`\n"
        md += f"* **Monitored Prompts Analyzed:** `{total_prompts}`\n"
        md += f"* **Distinct Winning Competitors:** `{total_comps}`\n"
        md += f"* **Top Recurring Gaps Identified:** `{len(insights)}`\n\n"
        md += "---\n\n## 📌 Executive Summary (High-Leverage Fixes)\n\n"
        md += (
            f"Across {total_prompts} monitored AI search prompts where `{brand_domain}` is losing citations, "
            f"the following recurring gaps represent the highest-leverage remediation actions. "
            f"Fixing these elements once on the core landing page will simultaneously improve visibility across multiple prompt clusters.\n\n"
        )

        md += "## 🎯 Ranked Recurring Fixes (Fix This Once, Win Multiple Prompts)\n\n"
        for idx, g in enumerate(insights, start=1):
            pct = round(100 * g.recurrence_count / g.total_prompts_analyzed) if g.total_prompts_analyzed else 0
            md += f"### #{idx} · {g.display_name} (`{g.citation_weight}` Priority)\n\n"
            md += f"* **Recurrence:** Appears on **{g.recurrence_count}/{g.total_prompts_analyzed} prompts ({pct}%)**\n"
            md += f"* **Priority Score:** `{g.priority_score}`\n"
            md += f"* **Affected Prompts:**\n"
            for p in g.affected_prompts:
                md += f"  - *\"{p}\"*\n"
            md += f"* **Seen on Competitor Pages:** {', '.join(f'[{u}]({u})' for u in g.example_competitor_urls)}\n"
            md += f"* **Action Required:** {g.representative_recommendation}\n\n"

        return md
